extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/graal/JulLogEventSubstitution.java

```diff
@@ -0,0 +1,39 @@
+package io.quarkus.logging.gelf.graal;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.LogRecord;
+
+import com.oracle.svm.core.annotate.Alias;
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+@TargetClass(className = "biz.paluch.logging.gelf.jul.JulLogEvent")
+public final class JulLogEventSubstitution {
+    @Alias
+    private static Map<String, String> threadNameCache;
+
+    @Substitute
+    private String getThreadName(LogRecord record) {
+        // the default implementation uses the ThreadMXBean to get the name of the thread corresponding to this thead ID
+        // SVM didn't support JMX so we use the Thread.getAllStackTraces().keySet(); instead
+        // an issue is openned upstream: https://github.com/mp911de/logstash-gelf/issues/213
```

Comment:
Have you tried this?

```java
if (record instanceof ExtLogRecord) { // from jboss-logmanager
   return ((ExtLogRecord)record).getThreadName();
}
```
I would expect `record` will *always* be an instance of `ExtLogRecord`; you should be able to assume that is always true.

Reply:
The substitution's code would certainly be accepted upstream, in this case I would remove it.
If we keep our substitution, I will try what you propose

Reply:
It would be nice to use the better code in JVM mode too, if possible, but having it just in the substitution should be OK for now I would think.

Reply:
I add a fast path in case the thread that log the message is the same that will handle it. Testing it in native this fast path is used.
I prefere to keep the substitution the same as the implementation that will be done upstream. Is it OK for you ?

Reply:
@dmlloyd I update the subsitution to the exact changes that will be done upstream. As I said there is a fast path to the current thread and it should be used in most cases.

Reply:
Looking at the GELF code, I think that none of the classes are really appropriate for Quarkus.  It seems like there should be a custom handler that includes proper support for our extended fields.

I see that `JBoss7JulLogEvent` exists but that's pretty old and doesn't account for the extra fields.  I think we want to just use a custom handler with a custom log event type instead of relying on substitutions for this.

Reply:
As I said, the substitution is temporary as it is already merged upstream.

Maybe the best is to use the [JBoss7GelfLogHandler](https://github.com/mp911de/logstash-gelf/blob/master/src/main/java/biz/paluch/logging/gelf/jboss7/JBoss7GelfLogHandler.java) that is designed to be used with jboss-logmanager and send a PR to the `logstash-gelf` library if we want to provides some features/fixes like gettting the thread name from the log event.

Regarding extended field, do you mean MDC and additionale filelds? It is supported inside logstahs-gelf but not yet inside this extension.

Reply:
No, I mean the thread ID and name.  It's all in the log record already; the rest of this code should not be needed in Quarkus (or in WildFly for that matter).

Reply:
So the best is to use the JBoss7GelfLogHandler and send a PR to `logstash-gelf` to override the `getThreadName()` method right ?

So we don't need to maintain our own GelfLogHandler implementation and others that uses `logstash-gelf` will also take advantage of this optimization!

Reply:
Yeah something like that.  They also have a WildFly one.  I think JBoss 7 used a version of LogManager from before when we were recording the thread name & ID though, so the WildFly one might be better to extend.  There could be a WildFly record which extends the JBoss7 one with the new stuff.

Reply:
The WildFly and the JBoss7 ones seems to be very identical, they both use the ExtLogRegord so if one didn't have access to the thread name the other will not have neither.

Maybe we need to upstream an upgrade of the `jboss-logmanager` library.

I will have a look later, native is not working again I need to understand why :(

Reply:
> The WildFly and the JBoss7 ones seems to be very identical, they both use the ExtLogRegord so if one didn't have access to the thread name the other will not have neither.

`ExtLogRecord` got the thread name field in version 1.1.0.  But looking back, the oldest log manager version ever used by JBoss AS 7 was 1.2.0, so I think we don't need to worry about it.  So I will pivot and suggest simply that `JBoss7JulLogEvent` be updated upstream to override `getThreadName` as I suggested earlier.

Reply:
done